## Análisis de datos - Clase 6

---
### Técnicas de selección de features
---

In [169]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.datasets import fetch_openml
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import MinMaxScaler, OrdinalEncoder, StandardScaler
from category_encoders import BinaryEncoder, OneHotEncoder
from sklearn.feature_selection import VarianceThreshold, SelectKBest,SelectFpr, chi2, f_classif, mutual_info_classif
from sklearn.model_selection import train_test_split


#### Para este ejemplo usamos el dataset 'adult' que tiene información capturada durante un censo. El objetivo es, dado un conjunto de características de adultos, predecir si su rango salarial estará por encima de 50K (problema de clasificación binaria).

In [170]:
# Cargar el dataset de censo de adultos
adultos = fetch_openml(name='adult', version=2, as_frame=True)
X = adultos.data.copy()
y = adultos.target.copy()

#### Mini-EDA

In [171]:
X.head()

,age,workclass,fnlwgt,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loss,hours-per-week,native-country
0,25,Private,226802,11th,7,Never-married,Machine-op-inspct,Own-child,Black,Male,0,0,40,United-States
1,38,Private,89814,HS-grad,9,Married-civ-spouse,Farming-fishing,Husband,White,Male,0,0,50,United-States
2,28,Local-gov,336951,Assoc-acdm,12,Married-civ-spouse,Protective-serv,Husband,White,Male,0,0,40,United-States
3,44,Private,160323,Some-college,10,Married-civ-spouse,Machine-op-inspct,Husband,Black,Male,7688,0,40,United-States
4,18,NaN,103497,Some-college,10,Never-married,NaN,Own-child,White,Female,0,0,30,United-States


In [172]:
X.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48842 entries, 0 to 48841
Data columns (total 14 columns):
 #   Column          Non-Null Count  Dtype   
---  ------          --------------  -----   
 0   age             48842 non-null  int64   
 1   workclass       46043 non-null  category
 2   fnlwgt          48842 non-null  int64   
 3   education       48842 non-null  category
 4   education-num   48842 non-null  int64   
 5   marital-status  48842 non-null  category
 6   occupation      46033 non-null  category
 7   relationship    48842 non-null  category
 8   race            48842 non-null  category
 9   sex             48842 non-null  category
 10  capital-gain    48842 non-null  int64   
 11  capital-loss    48842 non-null  int64   
 12  hours-per-week  48842 non-null  int64   
 13  native-country  47985 non-null  category
dtypes: category(8), int64(6)
memory usage: 2.6 MB


In [173]:
X.isna().mean().round(4)*100

age               0.00
workclass         5.73
fnlwgt            0.00
education         0.00
education-num     0.00
marital-status    0.00
occupation        5.75
relationship      0.00
race              0.00
sex               0.00
capital-gain      0.00
capital-loss      0.00
hours-per-week    0.00
native-country    1.75
dtype: float64

In [174]:
y.head()

0    <=50K
1    <=50K
2     >50K
3     >50K
4    <=50K
Name: class, dtype: category
Categories (2, object): ['<=50K', '>50K']

In [ ]:
# Definir características
num_features = ['age', 'fnlwgt', 'capital-gain', 'capital-loss', 'hours-per-week']
cat_nom_features = ['workclass', 'marital-status', 'occupation', 'relationship', 'race', 'sex', 'native-country']
cat_ord_features = ['education-num'] # Nivel de educación codificado numéricamente


In [177]:
X.drop(columns='education', inplace=True) # Elimino "education" porque ya tengo "education-num"

# Verifico
X.columns.to_list()

['age',
 'workclass',
 'fnlwgt',
 'education-num',
 'marital-status',
 'occupation',
 'relationship',
 'race',
 'sex',
 'capital-gain',
 'capital-loss',
 'hours-per-week',
 'native-country']

---
#### Paso 1: Dividir en train y test
---

In [178]:

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
print("Tamaño train:", X_train.shape, "Test:", X_test.shape)

Tamaño train: (39073, 13) Test: (9769, 13)


---
#### Paso 2: Imputación
---

In [179]:
# Antes de imputar:
X_train.isna().mean().round(4)*100

age               0.00
workclass         5.79
fnlwgt            0.00
education-num     0.00
marital-status    0.00
occupation        5.82
relationship      0.00
race              0.00
sex               0.00
capital-gain      0.00
capital-loss      0.00
hours-per-week    0.00
native-country    1.72
dtype: float64

In [180]:
X_test.isna().mean().round(4)*100

age               0.00
workclass         5.48
fnlwgt            0.00
education-num     0.00
marital-status    0.00
occupation        5.49
relationship      0.00
race              0.00
sex               0.00
capital-gain      0.00
capital-loss      0.00
hours-per-week    0.00
native-country    1.90
dtype: float64

In [181]:
# Imputar categóricas con moda (fit en train, transformar ambos)
# Solo imputo las categóricas nominales que son las únicas que tienen faltantes
cat_imputer = SimpleImputer(strategy='most_frequent')
X_train[cat_nom_features] = cat_imputer.fit_transform(X_train[cat_nom_features])

# No hay faltantes en test, pero lo aplico por consistencia y best-practice
X_test[cat_nom_features] = cat_imputer.transform(X_test[cat_nom_features])


In [182]:
# Verifico que no quedaron más faltantes en train:
X_train.isna().mean().round(4)*100

age               0.0
workclass         0.0
fnlwgt            0.0
education-num     0.0
marital-status    0.0
occupation        0.0
relationship      0.0
race              0.0
sex               0.0
capital-gain      0.0
capital-loss      0.0
hours-per-week    0.0
native-country    0.0
dtype: float64

---
#### Paso 3: Codificación
---

In [183]:
# Antes de codificar:
X_train[cat_nom_features].head()

,workclass,marital-status,occupation,relationship,race,sex,native-country
37193,Private,Married-civ-spouse,Machine-op-inspct,Husband,White,Male,United-States
31093,Self-emp-not-inc,Married-civ-spouse,Craft-repair,Husband,White,Male,United-States
33814,Local-gov,Married-civ-spouse,Craft-repair,Husband,White,Male,United-States
14500,Private,Never-married,Adm-clerical,Not-in-family,White,Female,United-States
23399,Private,Never-married,Transport-moving,Not-in-family,White,Male,United-States


In [184]:
X_train[cat_nom_features].shape

(39073, 7)

In [185]:
# Ojo que la variable target también hay que transformarla de alguna manera (es binaria)
y_train.head()

37193    <=50K
31093    <=50K
33814    <=50K
14500    <=50K
23399    <=50K
Name: class, dtype: category
Categories (2, object): ['<=50K', '>50K']

In [186]:
# Variable target (y) con OrdinalEncoder
oe_target = OrdinalEncoder(categories=[['<=50K', '>50K']])
y_train_encoded = oe_target.fit_transform(y_train.to_frame()).ravel()
y_test_encoded = oe_target.transform(y_test.to_frame()).ravel()


ohe = OneHotEncoder(use_cat_names=True)
X_train_cat = ohe.fit_transform(X_train[cat_nom_features])

# Guardamos el nombre de las nuevas columnas generadas por el codificador
nombres_cols_codificadas = columns=ohe.get_feature_names_out(cat_nom_features)

X_train_cat = ohe.fit_transform(X_train[cat_nom_features])
X_test_cat = ohe.transform(X_test[cat_nom_features])



In [187]:
X_train_cat.shape

(39073, 83)

In [188]:
X_train_cat.head()

,workclass_Private,workclass_Self-emp-not-inc,workclass_Local-gov,workclass_Federal-gov,workclass_Self-emp-inc,workclass_State-gov,workclass_Without-pay,workclass_Never-worked,marital-status_Married-civ-spouse,marital-status_Never-married,...,native-country_Vietnam,native-country_Haiti,native-country_Outlying-US(Guam-USVI-etc),native-country_Peru,native-country_Cambodia,native-country_Greece,native-country_Nicaragua,native-country_Thailand,native-country_Laos,native-country_Holand-Netherlands
37193,1,0,0,0,0,0,0,0,1,0,...,0,0,0,0,0,0,0,0,0,0
31093,0,1,0,0,0,0,0,0,1,0,...,0,0,0,0,0,0,0,0,0,0
33814,0,0,1,0,0,0,0,0,1,0,...,0,0,0,0,0,0,0,0,0,0
14500,1,0,0,0,0,0,0,0,0,1,...,0,0,0,0,0,0,0,0,0,0
23399,1,0,0,0,0,0,0,0,0,1,...,0,0,0,0,0,0,0,0,0,0


In [189]:
y_train_encoded

array([0., 0., 0., ..., 0., 0., 1.])

---
#### Paso 4: Normalización 
---

In [190]:
scaler = MinMaxScaler()

X_train_a_escalar = pd.concat([X_train[num_features], X_train[cat_ord_features]], axis=1)
X_test_a_escalar = pd.concat([X_test[num_features], X_test[cat_ord_features]], axis=1)

cols = num_features + cat_ord_features # Para mantener el mismo orden de columnas después de escalar

# Armo los DF con los datos escalados
X_train_num_escalado = pd.DataFrame(scaler.fit_transform(X_train_a_escalar), columns=cols, index=X_train.index)
X_test_num_escalado = pd.DataFrame(scaler.transform(X_test_a_escalar), columns=cols, index=X_test.index)

# Rearmamos con las col categóricas codificadas binarias y las numéricas+ ord escaladas
X_train_final = pd.concat([X_train_num_escalado, X_train_cat], axis=1).reset_index(drop=True)
X_test_final = pd.concat([X_test_num_escalado, X_test_cat], axis=1).reset_index(drop=True)


In [191]:
X_train_final.head()

,age,fnlwgt,capital-gain,capital-loss,hours-per-week,education-num,workclass_Private,workclass_Self-emp-not-inc,workclass_Local-gov,workclass_Federal-gov,...,native-country_Vietnam,native-country_Haiti,native-country_Outlying-US(Guam-USVI-etc),native-country_Peru,native-country_Cambodia,native-country_Greece,native-country_Nicaragua,native-country_Thailand,native-country_Laos,native-country_Holand-Netherlands
0,0.342466,0.090253,0.0,0.0,0.397959,0.600000,1,0,0,0,...,0,0,0,0,0,0,0,0,0,0
1,0.479452,0.110528,0.0,0.0,0.346939,0.333333,0,1,0,0,...,0,0,0,0,0,0,0,0,0,0
2,0.232877,0.108922,0.0,0.0,0.397959,0.533333,0,0,1,0,...,0,0,0,0,0,0,0,0,0,0
3,0.150685,0.077088,0.0,0.0,0.397959,0.533333,1,0,0,0,...,0,0,0,0,0,0,0,0,0,0
4,0.397260,0.010957,0.0,0.0,0.500000,0.666667,1,0,0,0,...,0,0,0,0,0,0,0,0,0,0


In [192]:
X_test_final.head()

,age,fnlwgt,capital-gain,capital-loss,hours-per-week,education-num,workclass_Private,workclass_Self-emp-not-inc,workclass_Local-gov,workclass_Federal-gov,...,native-country_Vietnam,native-country_Haiti,native-country_Outlying-US(Guam-USVI-etc),native-country_Peru,native-country_Cambodia,native-country_Greece,native-country_Nicaragua,native-country_Thailand,native-country_Laos,native-country_Holand-Netherlands
0,0.534247,0.014147,0.000000,0.0,0.397959,0.533333,1,0,0,0,...,0,0,0,0,0,0,0,0,0,0
1,0.109589,0.068297,0.000000,0.0,0.397959,0.533333,1,0,0,0,...,0,0,0,0,0,0,0,0,0,0
2,0.356164,0.107469,0.143441,0.0,0.397959,0.800000,1,0,0,0,...,0,0,0,0,0,0,0,0,0,0
3,0.205479,0.118514,0.000000,0.0,0.397959,0.533333,1,0,0,0,...,0,0,0,0,0,0,0,0,0,0
4,0.301370,0.180384,0.000000,0.0,0.295918,0.800000,1,0,0,0,...,0,0,0,0,0,0,0,0,0,0


---
#### Ahora el dataset está listo para aplicar filtros y seleccionar features.
---

#### 1. Correlación 

In [193]:
# Datos para la correlación de Pearson
X_train_num_corr = X_train_num_escalado[num_features].copy() # Solo numéricas para Pearson
y_train_corr = pd.Series(y_train_encoded, name='target') # lo convierto porque es un array

# Calculamos Pearsoncon la v. target
corr_con_target = (X_train_num_corr
                    .corrwith(y_train_corr, method='pearson')
                    .sort_values(ascending=False))

# Imprimir

print ("Correlación con la variable target (Pearson):")
corr_con_target

Correlación con la variable target (Pearson):


capital-loss      0.005831
age               0.004753
fnlwgt           -0.003723
capital-gain     -0.005208
hours-per-week   -0.008422
dtype: float64

In [194]:
# Datos para la correlación de Spearman
X_train_num_sp = X_train_num_escalado.copy() # Le puedo pasar una ordinal también
y_train_sp = pd.Series(y_train_encoded, name='target') # lo convierto porque es un array

# Calculamos Pearsoncon la v. target
sp_con_target = (X_train_num_sp
                    .corrwith(y_train_sp, method='spearman')
                    .sort_values(ascending=False))

print ("Correlación con la variable target (Spearman):")

# Imprimir
sp_con_target

Correlación con la variable target (Spearman):


education-num     0.010184
capital-loss      0.005340
age               0.004204
fnlwgt           -0.004319
hours-per-week   -0.007427
capital-gain     -0.007579
dtype: float64

In [195]:
# Datos para la correlación de Kendall
X_train_num_kdl = X_train_num_escalado.copy() # Le puedo pasar una ordinal también
y_train_kdl = pd.Series(y_train_encoded, name='target') # lo convierto porque es un array

# Calculamos Kendall con la v. target
kdl_con_target = (X_train_num_kdl
                    .corrwith(y_train_kdl, method='kendall')
                    .sort_values(ascending=False))

print ("Correlación con la variable target (Kendall):")

# Imprimir
kdl_con_target

Correlación con la variable target (Kendall):


education-num     0.009014
capital-loss      0.005280
age               0.003469
fnlwgt           -0.003527
hours-per-week   -0.006575
capital-gain     -0.007434
dtype: float64

Conclusión: no vemos indicios de relación lineal o monótona de las variables numéricas (y la ordinal) con la variable target. Esto no significa que no haya relación, sino que puede ser una relación más compleja que los métodos no pueden detectar.

#### 2. ANOVA

#### Analizamos el poder predictivo que tienen las variables numéricas sobre la variable target.



In [196]:
# ANOVA (numéricas, fit en train, transformar ambos)
selector_anova = SelectKBest(score_func=f_classif, k='all') # se puede usar f_regression cuando la v.a. target es numérica
X_train_num_best = selector_anova.fit_transform(X_train_final[num_features], y_train_encoded)

selected_indices = selector_anova.get_support() # Esta función me devuelve los features seleccionados
selected_features = X_train_final[num_features].columns[selected_indices]
p_values = selector_anova.pvalues_[selected_indices]
scores = selector_anova.scores_[selected_indices]

print(f"Numéricas antes de filtrar: {X_train_final[num_features].columns.tolist()}")
print("\nCaracterísticas seleccionadas (ANOVA):")
for feature, score in zip(selected_features, scores):
    print(f"{feature}: {score:.4f}")

print("\nP-valores de todas las características:")
for feature, p_valor in zip(X_train_final[num_features].columns, p_values):
    print(f"{feature}: {p_valor}")

# Aplicar el filtro a los datos de test:
X_test_num_best = selector_anova.transform(X_test_final[num_features])


Numéricas antes de filtrar: ['age', 'fnlwgt', 'capital-gain', 'capital-loss', 'hours-per-week']

Características seleccionadas (ANOVA):
age: 2241.2623
fnlwgt: 1.0314
capital-gain: 2017.2611
capital-loss: 909.6770
hours-per-week: 2119.3098

P-valores de todas las características:
age: 0.0
fnlwgt: 0.30983561615579164
capital-gain: 0.0
capital-loss: 1.4375880057715982e-197
hours-per-week: 0.0


Conclusión: el p-valor resultante de calcular Anova entre la variable fnlwgt es >0.05 y por lo tanto no hay evidencia de que este feature permita "separar" bien las clases de la variable target.
De las otras variables analizadas, se observa que age y hours-per-week parecen ser predictores fuertes.


#### 3. Chi-cuadrado


#### Identificar la relación entre las categóricas de entrada y la variable categórica de salida 

In [197]:
X_train_final

,age,fnlwgt,capital-gain,capital-loss,hours-per-week,education-num,workclass_Private,workclass_Self-emp-not-inc,workclass_Local-gov,workclass_Federal-gov,...,native-country_Vietnam,native-country_Haiti,native-country_Outlying-US(Guam-USVI-etc),native-country_Peru,native-country_Cambodia,native-country_Greece,native-country_Nicaragua,native-country_Thailand,native-country_Laos,native-country_Holand-Netherlands
0,0.342466,0.090253,0.0,0.0,0.397959,0.600000,1,0,0,0,...,0,0,0,0,0,0,0,0,0,0
1,0.479452,0.110528,0.0,0.0,0.346939,0.333333,0,1,0,0,...,0,0,0,0,0,0,0,0,0,0
2,0.232877,0.108922,0.0,0.0,0.397959,0.533333,0,0,1,0,...,0,0,0,0,0,0,0,0,0,0
3,0.150685,0.077088,0.0,0.0,0.397959,0.533333,1,0,0,0,...,0,0,0,0,0,0,0,0,0,0
4,0.397260,0.010957,0.0,0.0,0.500000,0.666667,1,0,0,0,...,0,0,0,0,0,0,0,0,0,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
39068,0.260274,0.423539,0.0,0.0,0.397959,0.533333,1,0,0,0,...,0,0,0,0,0,0,0,0,0,0
39069,0.232877,0.064750,0.0,0.0,0.500000,0.600000,1,0,0,0,...,0,0,0,0,0,0,0,0,0,0
39070,0.150685,0.161537,0.0,0.0,0.397959,0.600000,1,0,0,0,...,0,0,0,0,0,0,0,0,0,0
39071,0.397260,0.057298,0.0,0.0,0.397959,0.666667,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0


In [198]:
# Chi-Square (categóricas, fit en train, transformar ambos)
selector_chi2 = SelectKBest(score_func=chi2, k=10)
X_train_cat_best = selector_chi2.fit_transform(X_train_final[nombres_cols_codificadas], y_train_encoded)

chi2_selected_indices = selector_chi2.get_support()
chi2_selected_features = X_train_final[nombres_cols_codificadas].columns[chi2_selected_indices]

scores = selector_chi2.scores_[chi2_selected_indices]
p_values = selector_chi2.pvalues_[chi2_selected_indices]

print(f"Categóricas antes de filtrar: {X_train_final[nombres_cols_codificadas].columns.tolist()}")

print("\nScores de los feature seleccionadas (Chi2):")
for feature, score in zip(chi2_selected_features, scores):
    print(f"{feature}: {score:.4f}")

print("\nP-valores de los features analizados:")
for feature, p_valor in zip(chi2_selected_features, p_values):
    print(f"{feature}: {p_valor}")

# Aplicar el filtro a los datos de test:
X_test_cat_best = selector_chi2.transform(X_test_final[nombres_cols_codificadas])

Categóricas antes de filtrar: ['workclass_Private', 'workclass_Self-emp-not-inc', 'workclass_Local-gov', 'workclass_Federal-gov', 'workclass_Self-emp-inc', 'workclass_State-gov', 'workclass_Without-pay', 'workclass_Never-worked', 'marital-status_Married-civ-spouse', 'marital-status_Never-married', 'marital-status_Divorced', 'marital-status_Separated', 'marital-status_Widowed', 'marital-status_Married-spouse-absent', 'marital-status_Married-AF-spouse', 'occupation_Machine-op-inspct', 'occupation_Craft-repair', 'occupation_Adm-clerical', 'occupation_Transport-moving', 'occupation_Other-service', 'occupation_Prof-specialty', 'occupation_Exec-managerial', 'occupation_Sales', 'occupation_Tech-support', 'occupation_Handlers-cleaners', 'occupation_Farming-fishing', 'occupation_Protective-serv', 'occupation_Priv-house-serv', 'occupation_Armed-Forces', 'relationship_Husband', 'relationship_Not-in-family', 'relationship_Other-relative', 'relationship_Own-child', 'relationship_Unmarried', 'relati

Conclusión: para los K=10 features filtrados, el p-valor es <0.05 y por lo tanto podemos decir que para esos 10 features categóricos considerados, la dependencia con la variable target es estadísticamente significativa (lo que vemos muy probablemente NO es una casualidad de nuestra muestra). 
Entre los 10 filtrados, podemos ver que:

| Posición | Feature                            | Chi² Score   | 
|------|------------------------------------|--------------|
| 1    | `marital-status_Married-civ-spouse` | 4183.99     | 
| 2    | `relationship_Husband`             | 3758.17     | 
| 3    | `marital-status_Never-married`     | 2687.14     | 
| 4    | `relationship_Own-child`           | 1700.35     | 
| 5    | `occupation_Exec-managerial`       | 1489.63     | 
| 6    | `sex_Female`                       | 1187.01     | 
| 7    | `relationship_Not-in-family`       | 1042.58     | 
| 8    | `occupation_Other-service`         | 845.01      | 
| 9    | `workclass_Self-emp-inc`           | 765.64      | 
| 10   | `relationship_Unmarried`           | 720.23      | 


#### 4. Información mutua


In [199]:
# Información mutua (todas, fit en train, transformar ambos)
selector_mi = SelectKBest(score_func=mutual_info_classif, k=20) # también se puede usar mutual_info_regression para problemas de ese tipo
X_train_best = selector_mi.fit_transform(X_train_final, y_train_encoded) # todos los tipos de features
selected_indices = selector_mi.get_support()
selected_features = X_train_final.columns[selected_indices]
scores = selector_mi.scores_[selected_indices]

print("\nCaracterísticas seleccionadas (Inf. mutua):")
for feature, score in zip(selected_features, scores):
    print(f"{feature}: {score:.4f}")

X_test_best = selector_mi.transform(X_test_final)


Características seleccionadas (Inf. mutua):
age: 0.0728
fnlwgt: 0.0327
capital-gain: 0.0814
capital-loss: 0.0358
hours-per-week: 0.0440
education-num: 0.0649
workclass_Private: 0.0143
workclass_Self-emp-inc: 0.0104
marital-status_Married-civ-spouse: 0.1055
marital-status_Never-married: 0.0659
marital-status_Divorced: 0.0110
occupation_Other-service: 0.0137
occupation_Exec-managerial: 0.0227
relationship_Husband: 0.0827
relationship_Not-in-family: 0.0208
relationship_Own-child: 0.0381
relationship_Unmarried: 0.0136
race_White: 0.0093
sex_Male: 0.0288
sex_Female: 0.0269


| Posición | Feature                              | Score de IM | 
|------|--------------------------------------|---------------------------|
| 1    | `marital-status_Married-civ-spouse` | 0.1069                   |
| 2    | `relationship_Husband`              | 0.0819                   | 
| 3    | `capital-gain`                      | 0.0799                   | 
| 4    | `age`                               | 0.0684                   | 
| 5    | `education-num`                     | 0.0669                   | 
| 6    | `marital-status_Never-married`      | 0.0649                   | 
| 7    | `relationship_Own-child`            | 0.0410                   | 
| 8    | `hours-per-week`                    | 0.0391                   | 
| 9    | `capital-loss`                      | 0.0360                   | 
| 10   | `fnlwgt`                            | 0.0320                   | 
| 11   | `sex_Male`                          | 0.0261                   |
| 12   | `sex_Female`                        | 0.0244                   | 
| 13   | `relationship_Not-in-family`        | 0.0209                   | 
| 14   | `occupation_Other-service`          | 0.0184                   | 
| 15   | `occupation_Exec-managerial`        | 0.0178                   | 
| 16   | `relationship_Unmarried`            | 0.0151                   | 
| 17   | `workclass_Private`                 | 0.0135                   | 
| 18   | `workclass_Self-emp-inc`            | 0.0098                   | 
| 19   | `marital-status_Separated`          | 0.0094                   | 
| 20   | `race_White`                        | 0.0082                   | 

IM nos dice que el estado civil casado con esposo/a civil es uno de los predictores más fuertes. Coincide con el hecho de que la variable relationship_husband aparezca en el top....
Otra cosa interesante es que las variables numéricas como edad, capital-gain, etc. muestran fuerte poder predictivo, pero evidentemente la relación con la va. target no es lineal o monótona y los métodos de correlación no la capturan. Por esto es muy importante no quedarse con un solo método!

### 5- Resumen de resultados 

[Link para hacer análisis en clase](https://docs.google.com/spreadsheets/d/198GnbIemMHQLya7O8YWnkGBmUuyb_cRM8Y9Jn4qKMWw/edit?gid=759025845#gid=759025845)

| Feature | ANOVA F-score | ANOVA p-valor| Pearson | Spearman | Kendall | Chi² score | Chi² p-valor | Info. Mutua |
|---|---|---|---|---|---|---|---|---|
| `age` | 2241.2623 | 0.0 | 0.004753 | 0.004204 | 0.003469 | — | — | 0.0684 |
| `fnlwgt` | 1.0314 | 0.3098 | -0.003723 | -0.004319 | -0.003527 | — | — | 0.0320 |
| `capital-gain` | 2017.2611 | 0.0 | -0.005208 | -0.007579 | -0.007434 | — | — | 0.0799 |
| `capital-loss` | 909.6770 | 1.44e-197 | 0.005831 | 0.005340 | 0.005280 | — | — | 0.0360 |
| `hours-per-week` | 2119.3098 | 0.0 | -0.008422 | -0.007427 | -0.006575 | — | — | 0.0391 |
| `education-num` | — | — | — | 0.010184 | 0.009014 | — | — | 0.0669 |
| `workclass_Private` | — | — | — | — | — | — | — | 0.0135 |
| `workclass_Self-emp-inc` | — | — | — | — | — | 765.6398 | 1.60e-168 | 0.0098 |
| `marital-status_Married-civ-spouse` | — | — | — | — | — | 4183.9877 | 0.0 | 0.1069 |
| `marital-status_Never-married` | — | — | — | — | — | 2687.1369 | 0.0 | 0.0649 |
| `marital-status_Separated` | — | — | — | — | — | — | — | 0.0094 |
| `occupation_Other-service` | — | — | — | — | — | 845.0124 | 8.83e-186 | 0.0184 |
| `occupation_Exec-managerial` | — | — | — | — | — | 1489.6329 | 0.0 | 0.0178 |
| `relationship_Husband` | — | — | — | — | — | 3758.1692 | 0.0 | 0.0819 |
| `relationship_Not-in-family` | — | — | — | — | — | 1042.5836 | 9.96e-229 | 0.0209 |
| `relationship_Own-child` | — | — | — | — | — | 1700.3521 | 0.0 | 0.0410 |
| `relationship_Unmarried` | — | — | — | — | — | 720.2276 | 1.19e-158 | 0.0151 |
| `race_White` | — | — | — | — | — | — | — | 0.0082 |
| `sex_Male` | — | — | — | — | — | — | — | 0.0261 |
| `sex_Female` | — | — | — | — | — | 1187.0050 | 4.07e-260 | 0.0244 |

#### Algo para pensar...si usamos estos datos, ¿estaremos introduciendo un bias/sesgo histórico en el modelo?
